WcaOnRails/app/controllers/competitions_controller.rb

```diff
@@ -87,19 +87,41 @@ def index
     end
 
     if @display == 'admin'
-      @competitions = @competitions.includes(:delegates, :delegate_report)
+      @competitions = @competitions.includes(:delegate_report)
     end
 
     unless params[:region] == "all"
-      @competitions = @competitions.select { |competition| competition.belongs_to_region?(params[:region]) }
+      @competitions = @competitions.where("countryId = ? or continentId = ?", params[:region], params[:region])
     end
 
     if params[:search].present?
-      @competitions = @competitions.select { |competition| competition.contains?(params[:search]) }
+      params[:search].split.all? do |part|
+        @competitions = @competitions.where(
+          "Competitions.name like ? or
+          Competitions.cityName like ? or
+          users.name like ? or
+          Competitions.venue like ? or
+          Competitions.cellName like ? or
+          Competitions.countryId like ?",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+          "%#{part}%",
+        )
+      end
     end
 
     unless params[:event_ids].empty?
-      @competitions = @competitions.select { |competition| competition.has_events_with_ids?(params[:event_ids]) }
+      params[:event_ids].each do |event_id|
+        ce_alias = "competition_events" + event_id
+        e_alias = "Events" + event_id
+        @competitions = @competitions.joins(
+          "join competition_events " + ce_alias + " ON " + ce_alias + ".competition_id = Competitions.id " \
+          "join Events " + e_alias + " ON " + e_alias + ".id = " + ce_alias + ".event_id",
+        ).where(e_alias + ".id = ?", event_id)
+      end
```

Comment:
I'd be interested in seeing a performance analysis between this way and the old way. I suspect we'll save some time just because mysql doesn't have to allocation ~3000 ruby objects for each query.

If we do decide to go this path (pushing the work to mysql), I'd like to see all this SQL logic brought into the Competition class, rather than living out in here in the model. It's a shame that we have to implement this logic twice (once for sql, and once for ruby), but at the very least they should live in the same place.